crates/sui-indexer-alt/src/pipeline/concurrent/watermark.rs

```diff
@@ -117,9 +123,17 @@ pub(super) fn watermark<H: Handler + 'static>(
                         .start_timer();
 
                     let mut watermark_needs_update = false;
-                    while let Some(pending) = precommitted.first() {
-                        match watermark.next_cmp(pending) {
+                    while let Some(pending) = precommitted.first_entry() {
+                        let part = pending.get();
+
+                        // Some rows from the next watermark have not landed yet.
+                        if !part.is_complete() {
```

Comment:
would be good to have some tests for the watermark `take` and `add` back and then finally `update` logic here, it's subtle and takes some time to follow the flow from indexed to a watermark gets committed.